### PR TITLE
Improve Gaussian precision

### PR DESCRIPTION
**Pull request by leeping**
This PR was prompted by an email that asked how tight we could possibly make the geometry optimization criteria.  The tightest I could manage for a water dimer at the HF/STO-3G level using Gaussian as the single point energy was 1e-11 for energy, 1e-10 for gradient rms/max, and 1e-8 for displacement.  To do this, I modified the Gaussian input/output codes to write 10 digits after the decimal instead of 7, when writing coordinates to the input file; and read the gradient from the formatted .chk file rather than the output file.  I also added an example.

I also got a report that the Q-Chem engine was failing with a "file `131.0` missing" error.  I implemented a fallback to use the gradient from the output file, but I don't know how to reproduce the problem yet.

### CODE REVIEW

**Comment by leeping** [comment]
@jthorton It would be great if you could take a quick look at this, since you wrote the original Gaussian input/output codes. Thanks!

File: examples/2-challenges/water2_gaussian_extreme_criteria/saved/2023-01-10/water2.log
Comment (jthorton) [inline examples/2-challenges/water2_gaussian_extreme_criteria/saved/2023-01-10/water2.log:162]: I notice that all of the energies reported from Gaussian here are the same to 10 dp despite the XYZ coords reported in the `water2_optim.xyz` being different. I also notice the energy change between steps once the energy converges seems to change between a few repeating values randomly. Iterations 25, 26, 28, 30 and 36 are `-1.990e-13` we also see iterations 27 and 31 have the same value which due to them being so small seem like noise from converging the SCF. Could it be that Gaussian only supports reading input coordinates to a certain precision or is it something else?

**Comment by jthorton** [comment]
Hi, @leeping thanks for tagging me, seems like a good idea to use the fchk files for the gradient not sure why I didn't in the first place. I did notice something a little strange with the example you attached that would be good to check before merging, otherwise LGTM!

File: examples/2-challenges/water2_gaussian_extreme_criteria/saved/2023-01-10/water2.log
Comment (leeping) [reply examples/2-challenges/water2_gaussian_extreme_criteria/saved/2023-01-10/water2.log:162]: Thanks, Josh - I think this is probably due to a precision limitation due to how Gaussian writes out the energy, or how the energy is stored as a double precision number.  In the `.fchk` file, the energy data looks like this: `-1.499412443593249E+02` - the smallest digit is on the order of 1e-13.  I think it's not printing out any more because the energy is stored as a double, which inherently has about 15-16 significant digits.

If the problem were solely due to the finite precision of the Gaussian output, I would expect energy changes to look like `1.000e-13, 2.000e-13` or something like that.  However, we're seeing energy differences that look like `-1.990e-13`.  I'm guessing this is because Python is storing the energy as a `float64`, which means when we take differences between such large numbers, the precision of the result is limited to only 1-2 bits in the double precision representation.  That could explain why the energy differences look somewhat "granular" in these final steps.

I agree the behavior seems odd but I don't think it's a solvable issue or a major issue.

**Comment by leeping** [comment]
Thanks, Josh! I just approved the above. If you could add an approving review I can merge it without the administrative bypass. :)

**Comment by jthorton** [comment]
I don't seem to have permission to add an official approving review happy to be added to the repo to provide one if you would like though :)

**Comment by leeping** [comment]
Ah, sorry about that.  I didn't realize that there were limits on who could submit approving reviews.

In the repo settings there was an option that limited "approving" privileges to those with explicit write permissions to the main repo.  I just turned off that option, so you should be able to submit an approving review now.  It seems a bit bureaucratic, but it's a good chance to figure out how all this works.

**Comment by leeping** [comment]
I also invited you as a collaborator with direct write access, but if you could check whether you can approve the PR before accepting the invite that'd be helpful.

**Review by jthorton — APPROVED** [review]
LGTM!

**Comment by jthorton** [comment]
Thanks, I can review with that setting changed without the collaborator access which is good to know!

**Comment by leeping** [comment]
Thanks! :)

**Comment by leeping** [comment]
Strangely, I'm still seeing "Review required".  I'm not sure where you clicked to approve changes, but was it similar to this screenshot?

image

I'm not allowed to approve it myself because PR authors aren't allowed to self-approve.  I still think it's overly bureaucratic and I might just change the settings in the future... :P

**Comment by jthorton** [comment]
Yes, I used that and noticed the tick on the review went grey rather than green, after accepting your collaboration invite I can see it has now gone green so that must have been the issue.

**Comment by leeping** [comment]
Thanks so much. I'll merge the PR now. :D

**Pull request merged**
